### PR TITLE
fix(dhcp): merge reservations for known MACs instead of failing

### PR DESCRIPTION
**Pull request by jpennin5**
## Description

**Problem:** `create_dhcp_reservation` fails with `api.err.MacUsed` for
any client that is actually on the network — the common case, not the
edge case.

**Root cause:** the controller already holds a `rest/user` record for any
client it has seen, and POSTing a new record for that MAC is rejected.

**Fix:** `create_dhcp_reservation` now looks the MAC up first and merges
into the existing record with PUT, matching case-insensitively, and falls
back to POST only for a genuinely unknown MAC.

**Relationship to #95:** two things this branch originally carried are
dropped in favour of #95, which landed first and took the opposite
position deliberately: create now reports `use_fixedip` as `True` when the
controller's response omits the key, on the grounds that create always
requests it, so an absent key is an echo omission rather than an inactive
reservation. That also keeps create's report consistent with
`list_dhcp_reservations`, which filters on the same flag. The test
asserting the old behaviour is removed rather than left contradicting
upstream's.

**Live verification (Network 10.5.67):** reserving an IP for an on-network
client previously failed with `api.err.MacUsed`; it now merges into the
existing record and the reservation takes effect.

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)

## Testing

- Tests cover the known-MAC merge (including case-insensitive matching),
  the unknown-MAC POST fallback, and the merged payload preserving
  existing record fields.
- `pytest tests/unit/` and `pre-commit run --all-files` pass.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (CHANGELOG.md)
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published (builds on merged #95)

## AI Assistance

This PR was created with assistance from Claude Code.

**Human Review Status:** ✅ Reviewed and approved by @jpennin5
**Testing:** Full unit suite passing; merge path verified against a live
Network 10.5.67 controller.
**Security Review:** Completed — all fixture values are synthetic or from
documentation ranges.

### CODE REVIEW

File: src/tools/dhcp_reservations.py [hunk 1]
```diff
@@ -205,7 +206,21 @@ async def create_dhcp_reservation(
...
+                break
+
+        if existing is not None:
+            merge = {k: v for k, v in payload.items() if k != "mac"}
+            response = await client.put(_endpoint(site_id, existing["_id"]), json_data=merge)
+        else:
+            response = await client.post(_endpoint(site_id), json_data=payload)
         items = _unwrap(response)
         if not items:
             raise APIError(f"DHCP reservation POST returned no data for {mac}")
```
[the remark this answers is not in the document]
Comment (jpennin5) [reply src/tools/dhcp_reservations.py]: Fixed in be57135: a partial user record without `_id` raises a clear APIError, and the no-data message names the method that actually ran (PUT or POST).

File: tests/unit/tools/test_dhcp_reservations_tools.py [hunk 2]
```diff
@@ -197,6 +197,38 @@ async def test_create_dry_run(self, local_settings: MagicMock) -> None:
...
+        Posting it would fail with api.err.MacUsed.
+        """
+        merged = {**sample_users[1], "fixed_ip": "192.168.10.200", "use_fixedip": True}
+        client = _mock_client({"data": sample_users})
+        client.put.return_value = {"data": [merged]}
+
+        with patch("src.tools.dhcp_reservations.UniFiClient") as MockClient:
+            MockClient.return_value = client
+            result = await dhcp.create_dhcp_reservation(
+                mac="aa:bb:cc:dd:ee:02",
```
[the remark this answers is not in the document]
Comment (jpennin5) [reply tests/unit/tools/test_dhcp_reservations_tools.py]: Fixed in be57135 — the merge test now passes an uppercase MAC against the lowercase fixture, so the case-insensitive match is actually exercised.

**Comment by enuno** [comment]
### Summary
Request changes. The PR is not yet merge-ready because the current status checks are failing.

### Critical findings
- Merge state: \ / \
- Failing check(s): Mermaid Diagram Sync Assistant
- The changed tooling surface should be revalidated after the failing check is resolved.

### Suggestions
1. Resolve the failing status check(s) and rerun CI.
2. Reconfirm the changed tooling behavior against the intended controller/fixture shapes.
3. Ensure regression tests cover the original failure mode and edge cases introduced here.
4. Update docs if any request/response or error semantics changed.

### Verification checklist for the author
- [ ] Required checks are passing
- [ ] Regression tests cover the changed path
- [ ] Lint / format / type checks pass locally
- [ ] Documentation is updated where user-facing behavior changed
- [ ] No secrets, credentials, or controller-identifying data were introduced

**Review by enuno — CHANGES_REQUESTED** [review]
Merge blocked until the failing status checks are resolved and the tooling changes are revalidated.

**Comment by jpennin5** [comment]
The failing check is the `Mermaid Diagram Sync Assistant` GitHub App, not a workflow in this repository. Its check-run output on this PR reads: "Webhook Processing Failed — An unexpected error occurred while processing the webhook. Please try pushing a new commit or check your repository configuration." That processing happens on the app's side; pushing new commits to these PRs has re-triggered the check with the same result, so it is not reachable from a fork.

This is the loop documented in #112: the app has errored on every fork PR since late June, fork CI runs sit at `action_required` awaiting maintainer approval, and the daily automated review then reads the app error as a failing required check and requests changes.

Every check that can conclude on a fork PR is green here (both Socket Security checks). The remaining steps are repo-side — approving the CI runs and either repairing/removing the app integration or unrequiring its check. #117 merged with the identical check state, so the merge path exists. Happy to rebase or adjust anything code-side the moment CI can actually run.

**Comment by enuno** [comment]
Summary: request changes

Critical findings:
- Merge conflict / dirty merge state blocks safe review.

Suggestions:
1. Rebase onto main and resolve conflicts in: CHANGELOG.md, src/tools/dhcp_reservations.py, tests/unit/tools/test_dhcp_reservations_tools.py.
2. Re-run the touched test suite and pre-commit.
3. Re-check the feature behavior called out in the PR description.

Verification checklist:
- [ ] Rebasing completed cleanly
- [ ] Tests pass
- [ ] Lint/type checks pass
- [ ] Documentation remains aligned with implementation

**Review by enuno — CHANGES_REQUESTED** [review]
Automated review: merge conflicts/dirtiness detected (CONFLICTING/DIRTY). Please rebase and resolve before resubmitting.

**Comment by jpennin5** [comment]
Rebased onto current `main` (b360022) — the conflict was in CHANGELOG, and it is resolved.

Resolution was gated rather than eyeballed: every CHANGELOG bullet already on `main` is verified still present (union, not overwrite), the sibling work merged into the shared files is verified intact, the full unit suite passes, and pre-commit passes for every file the PR touches.

Codecov's three lines are covered: the malformed-existing-record error and the unechoed-write error that names the verb used.

GitHub now reports this PR mergeable.

**Comment by enuno** [comment]
## Summary
Request changes.

## Critical findings
- CI is not green on this PR.
- Failing checks: Mermaid Diagram Sync Assistant; Pre-commit Hooks.
- Touched scope: CHANGELOG.md, src/tools, tests/unit.

## Suggestions
1. Fix the failing checks on the current branch and rerun them until the gate is green.
2. Re-verify the changed code paths and tests for the touched scope above.
3. If any failing security or lint checks are expected, document the rationale and follow-up needed.

## Verification checklist
- [ ] `pre-commit run --all-files` passes
- [ ] Security/lint checks are green or explicitly justified
- [ ] Relevant unit tests cover the changed behavior
- [ ] No unintended API or documentation regressions

**Review by enuno — CHANGES_REQUESTED** [review]
CI is red on this PR (Mermaid Diagram Sync Assistant; Pre-commit Hooks). Please resolve the failing checks before merge.
